{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Visualization of the KO+ChIP Gold Standard from:\n",
    "# Miraldi et al. (2018) \"Leveraging ATAC-seq data for transcriptional regulatory network inference in Th17 Cells\"\n",
    "\n",
    "# TO START: In the menu above, choose \"Cell\" --> \"Run All\", and network + heatmap will load\n",
    "# Change \"canvas\" to \"SVG\" (drop-down menu in cell below) to enable drag interactions with nodes & labels\n",
    "# More info about jp_gene_viz and user interface instructions are available on Github: \n",
    "# https://github.com/simonsfoundation/jp_gene_viz/blob/master/doc/dNetwork%20widget%20overview.ipynb\n",
    "# Info specific to the \"Multi-network\" view:\n",
    "# https://github.com/simonsfoundation/jp_gene_viz/blob/master/doc/Combined%20widgets.ipynb\n",
    "\n",
    "# directory containing gene expression data and network folder\n",
    "directory = \".\"\n",
    "# folder containing networks\n",
    "netPath = 'Networks'\n",
    "# name of gene expression file\n",
    "expressionFile = 'Th0_Th17_48hTh.txt'\n",
    "# sample condition for initial gene node color\n",
    "sampleConditionOfInt = 'Th17(48h)' \n",
    "\n",
    "# The starting conditions for each of the networks is a list of tuples.  Tuple entries are:\n",
    "# 0. network file name (column format) (as found in directory)\n",
    "# 1. column of the expression matrix that you want the nodes to be colored by\n",
    "# 2. network title, to which we'll add the gene and peak cutoffs\n",
    "# 3. cut off for edge strength, note TRN edges strengths are quantile for 15 TFs/gene, to see top 10 TFs/gene, \n",
    "#    increase cutoff to .33, etc.\n",
    "networkInits = [\n",
    "    ('ChIPsA17_KO75_sAaTh_bias50_combined_cut01_sp.tsv',sampleConditionOfInt,'ChIP/sA(Th17)+KO+sA(Th), TF mRNA',.33),\n",
    "    ('sAaTh_bias50_combined_cut01_sp.tsv',sampleConditionOfInt,'ChIP/sA(Th17)+KO+sA(Th), TFA',.33),\n",
    "    (\"KO75_KOrk_1norm_sp.tsv\",sampleConditionOfInt,'KO G.S. (26 TFs)',0),\n",
    "    (\"KC1p5_sp.tsv\",sampleConditionOfInt,'KO-ChIP G.S. (9 TFs)',0)]\n",
    "\n",
    "tfFocus = 1 # If 1, automatically applies the \"TF only\" function, so we can focus on TFs\n",
    "    # If 0, all genes shown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uncomment to run without install (in binder, for example)\n",
    "import sys\n",
    "if \"..\" not in sys.path:\n",
    "    sys.path.append(\"..\")\n",
    "from jp_gene_viz import dNetwork\n",
    "dNetwork.load_javascript_support()\n",
    "from jp_gene_viz import multiple_network\n",
    "from jp_gene_viz import LExpression\n",
    "LExpression.load_javascript_support()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./ChIPsA17_KO75_sAaTh_bias50_combined_cut01_sp.tsv\n",
      "('Reading network', './Networks/ChIPsA17_KO75_sAaTh_bias50_combined_cut01_sp.tsv')\n",
      "('Loading saved layout', './Networks/ChIPsA17_KO75_sAaTh_bias50_combined_cut01_sp.tsv.layout.json')\n",
      "Omitting edges, using canvas, and fast layout default because the network is large\n",
      "./sAaTh_bias50_combined_cut01_sp.tsv\n",
      "('Reading network', './Networks/sAaTh_bias50_combined_cut01_sp.tsv')\n",
      "('Loading saved layout', './Networks/sAaTh_bias50_combined_cut01_sp.tsv.layout.json')\n",
      "Omitting edges, using canvas, and fast layout default because the network is large\n",
      "./KO75_KOrk_1norm_sp.tsv\n",
      "('Reading network', './Networks/KO75_KOrk_1norm_sp.tsv')\n",
      "('Loading saved layout', './Networks/KO75_KOrk_1norm_sp.tsv.layout.json')\n",
      "Omitting edges, using canvas, and fast layout default because the network is large\n",
      "./KC1p5_sp.tsv\n",
      "('Reading network', './Networks/KC1p5_sp.tsv')\n",
      "('Loading saved layout', './Networks/KC1p5_sp.tsv.layout.json')\n",
      "Omitting edges, using canvas, and fast layout default because the network is large\n"
     ]
    }
   ],
   "source": [
    "networkList = list()  # this list will contain heatmap-linked network objects\n",
    "for networkInit in networkInits:\n",
    "    networkFile = networkInit[0]\n",
    "    curr = LExpression.LinkedExpressionNetwork()   \n",
    "    print directory + '/' + networkFile\n",
    "    curr.load_network(directory + '/' + netPath + '/' + networkFile)        \n",
    "    networkList.append(curr)     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c07869c33b874fef8d28e6069cdf5c94",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "A Jupyter Widget"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# visualize the networks -- HARD CODED for 4 networks:\n",
    "M = multiple_network.MultipleNetworks(\n",
    "    [[networkList[0], networkList[1]],\n",
    "    [networkList[2], networkList[3]]])\n",
    "M.svg_width = 500\n",
    "M.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/emiraldi/anaconda/lib/python2.7/site-packages/scipy/stats/stats.py:2238: RuntimeWarning: invalid value encountered in true_divide\n",
      "  np.expand_dims(sstd, axis=axis))\n"
     ]
    }
   ],
   "source": [
    "# Set network preferences\n",
    "count = 0\n",
    "for curr in networkList:\n",
    "    networkInit = networkInits[count]\n",
    "    # get title information + curr column for shading of figures\n",
    "    currCol = networkInit[1]\n",
    "    titleInf = networkInit[2]\n",
    "    threshhold = networkInit[3]\n",
    "\n",
    "    # set threshold\n",
    "    curr.network.threshhold_slider.value = threshhold\n",
    "    curr.network.apply_click(None)\n",
    "    curr.network.restore_click(None)\n",
    "    if tfFocus:\n",
    "        # focus on TF core    \n",
    "        curr.network.tf_only_click(None)\n",
    "        curr.network.layout_click(None)  \n",
    "        \n",
    "    # set title\n",
    "    curr.network.title_html.value = titleInf\n",
    "\n",
    "    # add labels\n",
    "    curr.network.labels_button.value=True  \n",
    "    curr.network.draw_click(None)\n",
    "\n",
    "    # Load heatmap\n",
    "    curr.load_heatmap(directory + '/' + expressionFile)\n",
    "    # color nodes according to a sample column in the gene expression matrix\n",
    "    curr.gene_click(None)       \n",
    "    curr.expression.transform_dropdown.value = 'Z score'\n",
    "    curr.expression.apply_transform()\n",
    "    curr.expression.col = currCol    \n",
    "    curr.condition_click(None)    \n",
    "    count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
